File: prompt_hr/prompt_hr/web_form/exit_questionnaire/exit_questionnaire.py
import frappe
import json
from prompt_hr.py.utils import get_hr_managers_by_company, send_notification_email


def get_context(context):
    # ? USED FOR WEB FORM PAGE CONTEXT (IF ANY)
    pass


# ? FUNCTION TO FETCH EXIT INTERVIEW QUESTIONS FOR AN EMPLOYEE
@frappe.whitelist()
def fetch_interview_questions(employee=None):
    user = frappe.session.user
    is_hr_or_admin = check_user_role_and_employee().get("is_hr_or_admin")
    # ? IF NOT HR/ADMIN, ONLY ALLOW ACCESS TO OWN QUESTIONS
    if not is_hr_or_admin:
        if not employee:
            employee = frappe.db.get_value("Employee", {"user_id": user}, "name")
        if employee != frappe.db.get_value("Employee", {"user_id": user}, "name"):
            frappe.throw("You are not authorized to fetch questions for another employee.")

    if not employee:
        frappe.throw("Employee not provided and no employee linked to the current user.")

    # ? FETCH THE QUIZ LINKED TO THIS EMPLOYEE'S EXIT INTERVIEW
    quiz = frappe.db.get_value("Exit Interview", {"employee": employee}, "custom_resignation_quiz")

    if not quiz:
        frappe.throw(f"No quiz found for employee {employee}.")

    # ? FETCH ALL QUESTIONS ASSOCIATED WITH THE QUIZ
    questions = frappe.get_all("LMS Quiz Question", filters={"parent": quiz}, fields=["question", "question_detail"])

    for question in questions:
        question_data  = frappe.get_all(
            "LMS Question",
            {"name": question.question, "type": "Open Ended", "custom_input_type": ["is", "set"]},
            ["custom_input_type", "custom_multi_checkselect_options"]
        )
        if question_data:
            question["input_type"] = question_data[0].custom_input_type
            if question_data[0].custom_multi_checkselect_options:
                question["multi_checkselect_options"] = question_data[0].custom_multi_checkselect_options.split("\n")
        else:
            question["input_type"] = None

    return questions


# ? FUNCTION TO SAVE RESPONSES FROM THE EXIT QUESTIONNAIRE WEB FORM
@frappe.whitelist()
def save_response(employee, response):
    # ? DESERIALIZE JSON STRING TO PYTHON OBJECT
    try:
        response = json.loads(response)
    except Exception as e:
        frappe.throw(f"Invalid response format: {e}")

    # ? CHECK PERMISSIONS: HR, ADMIN, OR SELF
    user = frappe.session.user
    roles = frappe.get_roles(user)
    is_hr_or_admin = check_user_role_and_employee().get("is_hr_or_admin")
    if not is_hr_or_admin:
        linked_emp = frappe.db.get_value("Employee", {"user_id": user}, "name")
        if employee != linked_emp:
            frappe.throw("You are not authorized to submit responses.")

    # ? FETCH EXIT INTERVIEW DOCUMENT
    exit_doc = frappe.get_all("Exit Interview", filters={"employee": employee}, fields=["name"])
    if not exit_doc:
        frappe.throw(f"Exit Interview not found for employee: {employee}")
    
    exit_doc_name = exit_doc[0].name
    doc = frappe.get_doc("Exit Interview", exit_doc_name)

    # ? CLEAR EXISTING RESPONSES
    doc.custom_questions = []

    # ? ADD EACH NEW RESPONSE
    for entry in response:
        question_id = entry.get("question")
        question_detail = frappe.db.get_value("LMS Question", question_id, "question")
        answer = entry.get("answer")
        if not question_id:
            continue

        doc.append("custom_questions", {
            "question": question_detail,
            "question_name": question_id,
            "answer": answer
        })

    doc.save(ignore_permissions=True)
    frappe.db.commit()

    # ? NOTIFY HR MANAGERS
    hr_managers = get_hr_managers_by_company(doc.company)
    send_notification_email(
        doctype="Exit Interview",
        docname=doc.name,
        recipients=hr_managers,
        notification_name="Exit Questionnaire Form Submission"
    )

    return {"status": "success", "message": "Responses saved successfully."}


# ? FUNCTION TO CHECK IF CURRENT USER IS HR/ADMIN AND GET LINKED EMPLOYEE IF NOT
@frappe.whitelist()
def check_user_role_and_employee():
    """
    CHECKS IF THE LOGGED-IN USER HAS HR OR ADMIN ROLES.
    IF NOT, RETURNS THE LINKED EMPLOYEE RECORD (IF ANY).

    RETURNS:
        dict: {
            "is_hr_or_admin": bool,  # True if user is HR/Admin
            "employee": str or None   # Employee name if user is not HR/Admin
        }
    """

    #! GET CURRENT LOGGED-IN USER
    user = frappe.session.user

    #! DEFINE ROLES THAT ARE CONSIDERED HR OR ADMIN
    HR_ADMIN_ROLES = {
        "S - HR Leave Approval",
        "S - HR leave Report",
        "S - HR L6",
        "S - HR L5",
        "S - HR L4",
        "S - HR L3",
        "S - HR L2",
        "S - HR L1",
        "S - HR Director (Global Admin)",
        "S - HR L2 Manager",
        "S - HR Supervisor (RM)",
        "System Manager"
    }

    #! FETCH ALL ROLES OF THE USER
    user_roles = set(frappe.get_roles(user))

    #? CHECK IF USER IS ADMIN OR HAS ANY HR/ADMIN ROLE
    is_hr_or_admin = user == "Administrator" or not HR_ADMIN_ROLES.isdisjoint(user_roles)

    employee = None
    #? IF USER IS NOT HR/ADMIN, GET LINKED EMPLOYEE
    if not is_hr_or_admin:
        employee = frappe.db.get_value("Employee", {"user_id": user}, "name")

    #! RETURN RESULT
    return {
        "is_hr_or_admin": is_hr_or_admin,
        "employee": employee
    }

File: prompt_hr/prompt_hr/web_form/exit_questionnaire/exit_questionnaire.js
$(document).ready(function () {

    // ? CALL THE BACKEND METHOD TO CHECK USER ROLE AND EMPLOYEE
    frappe.call({
        method: "prompt_hr.prompt_hr.web_form.exit_questionnaire.exit_questionnaire.check_user_role_and_employee",
        callback: function (response) {
            const isHrOrAdmin = response.message.is_hr_or_admin;
            const sessionEmployee = response.message.employee;

            if (isHrOrAdmin) {
                // ? ALLOW HR OR ADMIN TO FILL THE EMPLOYEE FIELD
                enableEmployeeField();
            } else {
                // ? SET THE EMPLOYEE FIELD TO SESSION'S EMPLOYEE
                setEmployeeField(sessionEmployee);
            }
        },
        error: function (err) {
            console.error("Error fetching user role and employee:", err);
        }
    });

    // ? FUNCTION TO ENABLE THE EMPLOYEE FIELD (ALLOW HR OR ADMIN TO SELECT)
    function enableEmployeeField() {
        let employeeField = getEmployeeField();
        if (employeeField.length > 0) {
            employeeField.prop('readonly', false);
        }
    }

    // ? FUNCTION TO SET THE EMPLOYEE FIELD VALUE FOR NON-HR/ADMIN (SET FROM SESSION)
    function setEmployeeField(employee) {
        let employeeField = getEmployeeField();
        if (employeeField.length > 0) {
            employeeField.val(employee).prop('readonly', true);
        }
    }

    // ? FUNCTION TO GET THE EMPLOYEE FIELD
    function getEmployeeField() {
        let field = $('select[name="employee"], input[name="employee"]');
        if (field.length === 0) {
            field = $('[data-fieldname="employee"]').find('input');
        }
        if (field.length === 0) {
            field = $('select.employee-field, input.employee-field');
        }
        return field;
    }

    // ? HIDE THE DEFAULT SAVE BUTTON BUT KEEP THE DISCARD BUTTON VISIBLE
    $('.right-area .submit-btn').hide();
    $('.web-form-actions .discard-btn').show();

    // ? PREVENT DEFAULT FORM SUBMISSION
    $('form.web-form').on('submit', function (e) {
        e.preventDefault();
        console.log("Default form submission prevented");
        return false;
    });

    // ? FUNCTION TO PERIODICALLY CHECK FOR CHANGES IN EMPLOYEE FIELD VALUE
    function setupValueChangeDetection() {
        let lastVal = '';
        let employeeField = getEmployeeField();

        if (employeeField.length > 0) {
            console.log("Setting up value change detection");
            lastVal = employeeField.val();

            // ? POLLING FOR VALUE CHANGES
            setInterval(function () {
                const currentVal = employeeField.val();
                if (currentVal !== lastVal && currentVal) {
                    console.log("Value changed from", lastVal, "to", currentVal);
                    lastVal = currentVal;
                    fetchQuestions(currentVal);
                }
            }, 500);

            // ? MANUAL EVENT HANDLERS
            employeeField.on('change input blur', function () {
                const val = $(this).val();
                console.log("Direct event detected, value:", val);
                if (val) {
                    fetchQuestions(val);
                }
            });

            // ? INITIAL VALUE TRIGGER
            const initialVal = employeeField.val();
            if (initialVal) {
                console.log("Initial value found:", initialVal);
                fetchQuestions(initialVal);
            }

            // ? MAKE EMPLOYEE FIELD READONLY AFTER INTERACTION
            employeeField.on('blur', function () {
                $(this).prop('readonly', true);
                console.log("Employee field set to read-only");
            });
        } else {
            setTimeout(setupValueChangeDetection, 1000);
        }
    }

    // ? BEGIN MONITORING EMPLOYEE FIELD
    setupValueChangeDetection();

    // ? FETCH INTERVIEW QUESTIONS FROM BACKEND
    function fetchQuestions(employee) {
        console.log("Fetching questions for employee:", employee);

        if ($('#question-container').length === 0) {
            $('form.web-form').append('<div id="question-container" class="mt-4"></div>');
        }

        $('#question-container').html('<p>Loading questions...</p>');

        frappe.call({
            method: "prompt_hr.prompt_hr.web_form.exit_questionnaire.exit_questionnaire.fetch_interview_questions",
            args: { employee: employee },
            callback: function (response) {
                if (response.message && response.message.length > 0) {
                    console.log("Questions fetched:", response.message);
                    displayQuestions(response.message);
                } else {
                    $('#question-container').html('<p>No questions available for this employee.</p>');
                }
            },
            error: function (err) {
                console.error("API error:", err);
                $('#question-container').html('<p>Error loading questions: ' + (err.message || 'Unknown error') + '</p>');
            }
        });
    }

    // ? RENDER THE QUESTIONS INTO THE DOM
    function displayQuestions(questions) {
        const questionContainer = $('#question-container');
        questionContainer.empty();

        // HEADER
        questionContainer.append(`
        <h3 class="mb-4 text-center fw-bold" style="color: #3b5bdb; font-family: 'Poppins', sans-serif;">
            Exit Interview Questions
        </h3>
        <hr style="border-top: 2px solid #3b5bdb;">
    `);

        questions.forEach((question, index) => {
            let questionHtml = '';

            // QUESTION CARD START
            questionHtml += `
            <div class="question-item mb-4 p-4 border rounded shadow-sm bg-light hover-shadow" style="font-size: 15px; position: relative;">
                <div class="d-flex align-items-center mb-3">
                    <span class="badge rounded-pill bg-primary text-white me-3 shadow-sm" style="margin-right:20px; font-size: 14px; min-width: 35px; text-align: center;">Q${index + 1}</span>
                    <label class="fw-semibold mb-0" style="font-size: 15px;">${question.question_detail || 'No question details available'}</label>
                </div>
        `;

            // INPUT FIELD
            if (question.input_type) {
                switch (question.input_type) {
                    case 'Checkbox':
                        if (Array.isArray(question.multi_checkselect_options)) {
                            question.multi_checkselect_options.forEach((opt, i) => {
                                questionHtml += `
                                <div class="form-check mb-2">
                                    <input class="form-check-input" type="checkbox" id="check-${index}-${i}" name="answer-${question.question}" value="${opt}">
                                    <label class="form-check-label" for="check-${index}-${i}" style="font-size:14px;">${opt}</label>
                                </div>
                            `;
                            });
                        }
                        break;

                    case 'Yes/No/NA':
                        questionHtml += `
                        <select class="form-select mb-2" name="answer-${question.question}" style="font-size: 14px; width: 100%; padding: 6px 12px;">
                            <option value="">Select</option>
                            <option value="Yes">Yes</option>
                            <option value="No">No</option>
                            <option value="NA">N/A</option>
                        </select>
                    `;
                        break;

                    case 'Single Line Input':
                        questionHtml += `
                        <input type="text" class="form-control mb-2" name="answer-${question.question}" style="font-size:14px;" placeholder="Enter your answer here ✍️">
                    `;
                        break;

                    case 'Dropdown':
                        if (Array.isArray(question.multi_checkselect_options)) {
                            questionHtml += `
                            <select class="form-select mb-2" name="answer-${question.question}" style="font-size: 14px; width: 100%; padding: 6px 12px;">
                                <option value="">Select an option</option>
                        `;
                            question.multi_checkselect_options.forEach(opt => {
                                questionHtml += `<option value="${opt}">${opt}</option>`;
                            });
                            questionHtml += `</select>`;
                        }
                        break;

                    case 'Date':
                        questionHtml += `
                        <input type="date" class="form-control mb-2" name="answer-${question.question}" style="font-size:14px;">
                    `;
                        break;

                    default:
                        questionHtml += `
                        <textarea class="form-control mb-2" rows="2" name="answer-${question.question}" style="font-size:14px;" placeholder="Your feedback..."></textarea>
                    `;
                }
            } else {
                questionHtml += `
                <textarea class="form-control mb-2" rows="2" name="answer-${question.question}" style="font-size:14px;" placeholder="Your feedback..."></textarea>
            `;
            }

            // QUESTION CARD END
            questionHtml += '</div>';

            // APPEND TO CONTAINER
            questionContainer.append(questionHtml);
        });

        // FOOTER NOTE
        questionContainer.append(`
        <p class="text-center text-muted fst-italic" style="margin-top: 20px;">Please answer all questions above.</p>
    `);

        addButtons();
        console.log("Rendered questions:", questions.length);
    }

    // ? ADD SAVE RESPONSE BUTTON (ONLY ONCE)
    function addButtons() {
        if ($('#save-response').length === 0) {
            const buttonsHtml = `
                <div class="mt-3">
                    <button type="button" id="save-response" class="btn btn-primary">Save Response</button>
                </div>
            `;
            $('form.web-form').append(buttonsHtml);

            $('#save-response').on('click', function (e) {
                e.preventDefault();
                e.stopPropagation();

                frappe.confirm(
                    'Are you sure you want to save your responses?',
                    function () {
                        saveResponse();
                    },
                    function () {
                        console.log("Save action cancelled.");
                    }
                );

                return false;
            });
        }
    }

    // ? COLLECT AND SAVE RESPONSES VIA API
    function saveResponse() {
        const responses = [];
        const employeeField = getEmployeeField();
        const employee = employeeField.val();

        //! COLLECT ANSWERS FROM ALL QUESTION TYPES
        $('#question-container')
            .find('input, textarea, select')
            .each(function () {
                const element = $(this);
                const nameAttr = element.attr('name');

                //? SKIP IF NO NAME ATTRIBUTE
                if (!nameAttr || !nameAttr.startsWith('answer-')) return;

                const questionName = nameAttr.replace('answer-', '');
                let answer = '';

                //? HANDLE CHECKBOXES (COLLECT MULTIPLE VALUES)
                if (element.attr('type') === 'checkbox') {
                    if (element.is(':checked')) {
                        // IF QUESTION ALREADY EXISTS, PUSH ADDITIONAL VALUE
                        const existing = responses.find(r => r.question === questionName);
                        if (existing) {
                            existing.answer += `, ${element.val()}`;
                        } else {
                            responses.push({
                                question: questionName,
                                answer: element.val()
                            });
                        }
                    }
                    return; // SKIP FURTHER PROCESSING FOR CHECKBOX
                }

                //? HANDLE OTHER INPUT TYPES
                answer = element.val();

                //? ONLY PUSH IF QUESTION NOT ALREADY ADDED (AVOID DUPLICATE CHECKBOX HANDLING)
                const existingResponse = responses.find(r => r.question === questionName);
                if (existingResponse) {
                    existingResponse.answer = answer;
                } else {
                    responses.push({
                        question: questionName,
                        answer: answer
                    });
                }
            });


        frappe.call({
            method: "prompt_hr.prompt_hr.web_form.exit_questionnaire.exit_questionnaire.save_response",
            args: {
                employee: employee,
                response: responses
            },
            callback: function (response) {
                console.log("Responses saved successfully:", response);
                showThankYouPage();
            },
            error: function (err) {
                console.error("Error saving responses:", err);
                alert("There was an error saving your responses: " + (err.message || 'Unknown error'));
            }
        });
    }

    // ? DISABLE ONLY THE DEFAULT SAVE BUTTON BUT KEEP DISCARD BUTTON
    setTimeout(function () {
        // ? TARGET ONLY THE SUBMIT/SAVE BUTTON IN WEB-FORM-ACTIONS, NOT ALL BUTTONS
        $('.web-form-actions .btn-primary, .right-area .submit-btn').off('click').prop('disabled', true);
        console.log("Disabled default save button, kept discard button functional");
    }, 1000);
});

// ? SHOW THANK YOU PAGE AFTER SUCCESSFUL RESPONSE SAVE
function showThankYouPage() {
    $('.web-form-container').html(`
        <div class="text-center py-5">
            <i class="fa fa-check-circle text-success" style="font-size: 48px;"></i>
            <h3 class="mt-3">Thank You!</h3>
            <p class="lead">Your information has been successfully updated.</p>
            <p>You may close this window now.</p>
        </div>
    `);
}
